source/extensions/filters/network/local_ratelimit/local_ratelimit.cc

```diff
@@ -10,26 +10,62 @@ namespace Extensions {
 namespace NetworkFilters {
 namespace LocalRateLimitFilter {
 
+SINGLETON_MANAGER_REGISTRATION(shared_local_ratelimit);
+
+LocalRateLimiterImplSharedPtr SharedRateLimitSingleton::get(
+    const envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit& proto_config,
+    std::function<LocalRateLimiterImplSharedPtr()> create_fn) {
+  ASSERT_IS_MAIN_OR_TEST_THREAD();
+
+  if (proto_config.share_key().empty()) {
+    // If there is no key, never share.
+    return create_fn();
+  }
+
+  const Key key{proto_config.share_key(), proto_config.token_bucket()};
+  auto it = limiters_.find(key);
+  if (it == limiters_.end()) {
+    ENVOY_LOG(debug, fmt::format("LocalRateLimit for share_key '{}' is creating a new token bucket "
+                                 "due to no match found",
+                                 proto_config.share_key()));
+    it = limiters_.emplace(key, create_fn()).first;
+  } else {
+    ENVOY_LOG(
+        debug,
+        fmt::format("LocalRateLimit for share_key '{}' is using an existing matching token bucket",
+                    proto_config.share_key()));
+  }
+  return it->second;
+}
+
 Config::Config(
     const envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit& proto_config,
-    Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime)
-    : rate_limiter_(Filters::Common::LocalRateLimit::LocalRateLimiterImpl(
-          std::chrono::milliseconds(
-              PROTOBUF_GET_MS_REQUIRED(proto_config.token_bucket(), fill_interval)),
-          proto_config.token_bucket().max_tokens(),
-          PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config.token_bucket(), tokens_per_fill, 1),
-          dispatcher,
-          Protobuf::RepeatedPtrField<
-              envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>())),
-      enabled_(proto_config.runtime_enabled(), runtime),
-      stats_(generateStats(proto_config.stat_prefix(), scope)) {}
+    Event::Dispatcher& dispatcher, Stats::Scope& scope, Runtime::Loader& runtime,
+    Singleton::Manager& singleton_manager)
+    : enabled_(proto_config.runtime_enabled(), runtime),
+      stats_(generateStats(proto_config.stat_prefix(), scope)),
+      shared_bucket_registry_(singleton_manager.getTyped<SharedRateLimitSingleton>(
```

Comment:
What will happen if the config get updated (e.g., by xDS) ? Will it reuse the singleton before update?

Reply:
The `Singleton::Manager` is owned by the `Server`, so it's lifetime is the same as the process, and objects are held by `weak_ptr`. So as long as a `local_ratelimit` filter exists, this singleton instance will exist. The token buckets in this singleton instance are held by `shared_ptr`, so they exist as long as at least 1 filter is still using them.

If the configuration is updated for this filter (eg the rate configuration is changed, or the `share_key` is changed), a new instance will be created for that configuration.

So in summary, token buckets will continue to exist as long as they are still in use.

But this question made me realize that the singleton will never free old, unused instances from `limiters_`. I pushed a fix for that issue.